Add Electron Transport Capability

melekderman opened the pull request:
✅  (uploaded/inserted (almost) as is; I will check these later to make them compatible with the new data refactoring)
✅ The file `element.py` to `object_/`
✅ Electron-related items in `reaction.py` to `object_/reaction.py` (append or insert, not overwrite the entire file)
✅  The folder `physics/electron/` to `transport/physics/`
👩🏻‍💻 Electron-related items in `material.py` to `object_/material.py` (append or insert, not overwrite the entire file)

melekderman committed:
add changes made so far

ilhamv committed:
Merge branch 'dev' into dev-electron

ilhamv committed:
minor edit

ilhamv committed:
edit variable types

ilhamv committed:
add elements to simulation object

ilhamv committed:
minor edit

ilhamv committed:
add a data placeholder in object util

ilhamv committed:
revert back to using base type

ilhamv commented:
Thanks, Melek!

As we planned, you can go ahead and drop the remaining items (material.py, etc). Don't worry about the failing tests. They are supposed to fail, and then from there, we collaboratively fix and organize them here.

Some comments on some edits I made in my recent commits:
- I see that you are already aware of the type-hinting stuff needed for Numba mode and code factory. 👍🏽 
- Note that now we have `DistributionBase` and `DataBase`, described in `object_/distribution.py` and `object_/data.py`, respectively. (See the changes in `reaction.py`)
- Note that I added the `Element` class and `elements` list in the simulation object description in `object_/simulation.py`. This is needed to get the new object discoverable by the code factory and simulation.

melekderman committed:
fix: correct fundamental electron transport physics to match new refactoring

melekderman commented:
@ilhamv 

Thank you so much for your review! I refactored the `reactions.py`, `element.py`, and `physics/electron/native.py`. I also moved the `ISOTOPIC_ABUNDANCE` data to `object_/util.py`, as you suggested. Your feedback would be greatly appreciated. 🙂 The remaining work is to insert `material.py` functions and to add the `EDEP` tally and related functions.

ilhamv committed:
add electron to physics __init__

ilhamv committed:
add electron to physics interface

ilhamv committed:
minor edits for neutron physics

ilhamv committed:
correct electron physics __init__

ilhamv committed:
correct electron physics interface. Also delete electron_production_xs

ilhamv committed:
minor edits on electron native physics.

ilhamv committed:
minor adjustments to excitation and bremsstrahlung

ilhamv commented:
Have you made sure whether the scattering kinematics data for elastic scattering and ionization are provided in the laboratory or the center-of-mass frame?

ilhamv commented:
In Ionization we have

```
    primary_alive_after = True
    if E_out <= ELECTRON_CUTOFF_ENERGY:
        edep += E_out
        particle["E"] = 0.0
        particle["alive"] = False
        primary_alive_after = False

    if T_delta <= ELECTRON_CUTOFF_ENERGY:
        edep += T_delta
        return edep
```

I think if the delta electron got cut off, but the primary electron is still alive, we should not `return` as we still need to determine the primary electron direction.

melekderman commented:
> In ionization we have:
> 
> ```
>     # Deposit all energy if below binding energy
>     if E <= B:
>         edep = E
>         particle["alive"] = False
>         particle["E"] = 0.0
>         return edep
> ```
> 
> Isn't it impossible for a sampled subshell to have a binding energy that is larger than the electron energy? I'm wondering, if it does occur in the simulation, does that mean something is incorrect in the subshell cross-sections used for sampling?

Sure, that is not possible. It seems like all the extra checks I added while debugging to make sure energy is conserved are still in the code. I will remove them. Thank you!

melekderman commented:
> About ELECTRON_CUTOFF_ENERGY used in `collision()` in `mcdc/transport/physics/electron/native.py`.
> 
> 1. Want to make sure: So, an electron with energy below the threshold will be absorbed if it collides? Or, when the said electron is created, it will be absorbed right away (does not have the chance to move to a collision site)?
> 2. Should the threshold be element dependent?
> 
> ===
> 
> After learning more about this, I think it is best to make the cutoff adjustable per region. But we can do that later. But still, I think the cut-off check should not be done in the `collision()` function, unless we have a mechanism that reduces the electron energy while free-streaming.

I noticed that (I think I added an extra check again during debugging), I put a cutoff check both inside the collision and in the reactions. The one inside the collision must have been for debugging. And since the particle cannot actually travel that path, since the particle is absorbed immediately after it’s created in the reaction, it cannot reach the cutoff check in the collision routine. I'll delete it. Thank you!

melekderman committed:
add an example (needs to update) and refs

melekderman committed:
change elastic scattering helpers location

melekderman committed:
add slab example for Al

melekderman committed:
Merge branch 'CEMeNT-PSAAP:dev' into dev-electron

melekderman committed:
examples that Ilham uploaded

melekderman committed:
Merge branch 'dev-electron' of https://github.com/melekderman/MCDC into dev-electron

melekderman committed:
electron related mcdc_get and mcdc_set files

melekderman committed:
some minor fixes before proceeding

melekderman committed:
add element support to material.py

melekderman committed:
Merge branch 'dev' into dev-electron
